File: day6/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"unicode"
)

func readInput(path string) ([]string, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func splitBlocks(lines []string) [][][]rune {
	width := 0
	for _, l := range lines {
		if len(l) > width {
			width = len(l)
		}
	}

	grid := make([][]rune, len(lines))
	for i, l := range lines {
		row := []rune(l)
		for len(row) < width {
			row = append(row, ' ')
		}
		grid[i] = row
	}

	used := make([]bool, width)
	for c := 0; c < width; c++ {
		for r := range grid {
			if grid[r][c] != ' ' {
				used[c] = true
				break
			}
		}
	}

	var blocks [][][]rune
	start := -1

	for i := 0; i <= width; i++ {
		active := i < width && used[i]
		if active && start == -1 {
			start = i
		} else if !active && start != -1 {
			block := make([][]rune, len(grid))
			for r := range grid {
				block[r] = grid[r][start:i]
			}
			blocks = append(blocks, block)
			start = -1
		}
	}

	return blocks
}

func solveBlock(block [][]rune) int64 {
	h := len(block)
	w := len(block[0])

	var numbers []int64
	var op rune

	for c := w - 1; c >= 0; c-- {
		var digits []rune
		for r := 0; r < h; r++ {
			ch := block[r][c]
			if unicode.IsDigit(ch) {
				digits = append(digits, ch)
			} else if ch == '+' || ch == '*' {
				op = ch
			}
		}

		if len(digits) > 0 {
			n, _ := strconv.ParseInt(string(digits), 10, 64)
			numbers = append(numbers, n)
		}
	}

	if op == '+' {
		var sum int64
		for _, n := range numbers {
			sum += n
		}
		return sum
	}

	result := int64(1)
	for _, n := range numbers {
		result *= n
	}
	return result
}

func main() {
	lines, err := readInput("input.txt")
	if err != nil {
		panic(err)
	}

	blocks := splitBlocks(lines)

	var total int64
	for _, b := range blocks {
		total += solveBlock(b)
	}

	fmt.Println(total)
}

File: day6/part2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_LINES 2000
#define MAX_WIDTH 2000

char grid[MAX_LINES][MAX_WIDTH];
int rows = 0, width = 0;

long long solve_block(int start, int end) {
    long long nums[128];
    int count = 0;
    char op = 0;

    for (int c = end - 1; c >= start; c--) {
        char buf[64];
        int len = 0;

        for (int r = 0; r < rows; r++) {
            char ch = grid[r][c];
            if (isdigit(ch)) buf[len++] = ch;
            else if (ch == '+' || ch == '*') op = ch;
        }

        if (len > 0) {
            buf[len] = '\0';
            nums[count++] = atoll(buf);
        }
    }

    if (op == '+') {
        long long sum = 0;
        for (int i = 0; i < count; i++) sum += nums[i];
        return sum;
    }

    long long prod = 1;
    for (int i = 0; i < count; i++) prod *= nums[i];
    return prod;
}

int main(void) {
    FILE *f = fopen("input.txt", "r");
    if (!f) {
        perror("input.txt");
        return 1;
    }

    while (fgets(grid[rows], MAX_WIDTH, f)) {
        int len = strlen(grid[rows]);
        if (grid[rows][len - 1] == '\n') grid[rows][--len] = '\0';
        if (len > width) width = len;
        rows++;
    }
    fclose(f);

    for (int r = 0; r < rows; r++) {
        for (int c = strlen(grid[r]); c < width; c++) {
            grid[r][c] = ' ';
        }
    }

    int used[MAX_WIDTH] = {0};
    for (int c = 0; c < width; c++) {
        for (int r = 0; r < rows; r++) {
            if (grid[r][c] != ' ') {
                used[c] = 1;
                break;
            }
        }
    }

    long long total = 0;
    int start = -1;

    for (int c = 0; c <= width; c++) {
        int active = (c < width && used[c]);
        if (active && start == -1) start = c;
        else if (!active && start != -1) {
            total += solve_block(start, c);
            start = -1;
        }
    }

    printf("%lld\n", total);
    return 0;
}